### ISSUE TITLE
Projects built on .NET full won't run due to WebSocket4Net error

### ISSUE THREAD

**Issue by shawty** [skrusty/AsterNET.ARI#64]
Creating projects running on .NET core work perfectly, but to use speech synthesis I need to target the full .NET platform, as the speech synthesis assemblies are not yet available on core.

I've gone back in full .NET versions as far as I can without all the dependencies AsterNET uses, falling apart and the earliest version test is .NET v4.5.1

I have tested all the way up to v4.7.2 using both Visual Studio 2017 and Visual Studio 2019 (Release and Preview Builds) running as Admin and not as Admin.

In all cases, everything builds and references fine, but as soon as you try to run the app  the following exception is thrown.

![image](https://user-images.githubusercontent.com/1876454/59146760-03c36380-89ea-11e9-9c77-cc56fb4cb021.png)

For reference, I've also performed recommended NuGet updates (Taking Websocket up to 0.15.12) and as well as reinstalling against different versions, have also test built from scratch.

Output seen in image is 100% consistent every time, have tested this on 2 different asterisk servers and same result.

**Comment by shawty** [comment]
ADDITIONAL: The two different asterisk versions where the ones mentioned in my "404" report, and can both be connected to without issue using .NET core code

**Comment by shawty** [comment]
UPDATE: Cloned a copy of this repo, replaced the AsterNET.ARI and WebSocket4Net dlls installed using NuGet with a newly built AsterNet (Debug, under VS17) and WebSocket4Net provided in the build output from the cloned repo.

With just those 2 DLL's replaced, the full .NET app shown above works without issue.

**Comment by shawty** [comment]
UPDATE2: As previous, only this time, swapped out "SuperSocket.ClientEngine" , removing NuGet auto installed version and replacing with version installed in the locally compiled copy of the repo, and exception goes away.

**Comment by shawty** [comment]
UPDATE3: Same thing occurs with Newtonsoft JSON when a message is recieved by the app from Asterisk.

![image](https://user-images.githubusercontent.com/1876454/59147986-4f7e0900-89fa-11e9-8720-9c939ea48778.png)

I suspect, another swap will resolve.

**Comment by shawty** [comment]
UPDATE4: Newtonsoft Json aborts with the same error, when swapped out for the version installed by NuGet with the version from the cloned repo

![image](https://user-images.githubusercontent.com/1876454/59148023-ca472400-89fa-11e9-8ecc-2f666cf0da49.png)

**Comment by shawty** [comment]
For reference, currently building under VS2017 with full .NET 4.5.1 (Lowest common denominator)

![image](https://user-images.githubusercontent.com/1876454/59148036-fd89b300-89fa-11e9-9e5d-9dfd88e363e3.png)

Current package set is 

![image](https://user-images.githubusercontent.com/1876454/59148046-2316bc80-89fb-11e9-812a-a5251076e975.png)

AsterNET.ARI, Newtonsoft JSON, SuperSocket and WebSocket4net are all being loaded from debug build output in repo clone, the rest are auto installed from NuGet following the first install after the initial project creation.

**Comment by shawty** [comment]
UPDATE5: Created a new project from scratch, full .NET 4.5.1, console app, did not use NuGet to install AsterNET.ARI, instead referenced ONLY the AsterNET.ARI, NewtonsoftJson, SuperSocket and WebSocket4Net assemblies from the previously cloned and debug built repo (Built using VS2017 with no changes to your original project, IE: load and build) 

This has resulted in an EXE that works without throwing any exceptions, but, the actions requested of the code

a) Failed to play the sound (Error unsupported scheme)
b) Issued a 404 error (but not an exception) [As seen in my previous report] on hanging up the channel.

For reference, here is the code used:

```
using AsterNET.ARI;
using AsterNET.ARI.Models;
using System;

namespace netfulltestusingclone
{
  class Program
  {
    //Boiler plate start

    private static string ariServer = "192.168.17.207";
    private static int ariPort = 8088;
    private static string userName = "******";
    private static string password = "******";
    private static string applicationName = "pork";

    public static AriClient ariClient;
    public static StasisEndpoint endPoint;

    static void Main(string[] args)
    {
      endPoint = new StasisEndpoint(ariServer, ariPort, userName, password);
      ariClient = new AriClient(endPoint, applicationName);

      ariClient.OnStasisStartEvent += OnStasisStart;
      ariClient.OnStasisEndEvent += OnStasisEnd;

      ariClient.Connect();

      bool done = false;
      Console.WriteLine("Press * to stop service");
      while (!done)
      {
        var lastKey = Console.ReadKey();
        switch (lastKey.KeyChar.ToString())
        {
          case "*":
            done = true;
            break;
        }
      }

      ariClient.Disconnect();

    }

    static void OnStasisStart(object sender, StasisStartEvent e)
    {
      ariClient.Channels.Answer(e.Channel.Id);

      TestPlay(e.Channel);
    }

    static void OnStasisEnd(object sender, StasisEndEvent e)
    {
      ariClient.Channels.Hangup(e.Channel.Id, "normal");
    }

    //Boiler plate end

    static void TestPlay(Channel c)
    {
      ariClient.Channels.Play(c.Id, "1-for-am-2-for-pm");
    }

  }
}
```
The extension to call it is a simple 3 line

```
NoOp
Statis(pork)
Hangup()
```

(Yes I do use stupid names when I'm testing :-)  )

Output while running shows:

![image](https://user-images.githubusercontent.com/1876454/59148209-adf8b680-89fd-11e9-8a52-e54ddb044f7f.png)

**Comment by shawty** [comment]
PS: This is running against the Ubuntu 18.04 version of Asterisk which I confirmed to not have the 404 error in my previous issue report.

**Comment by shawty** [comment]
Sound scheme error is in fact not an error.  Sound uri's sent to asterisk need to be always qualified with a scheme in the Ubuntu 18.04 version.

The asterisk source code shows:

```
#define SOUND_URI_SCHEME "sound:"
#define RECORDING_URI_SCHEME "recording:"
#define NUMBER_URI_SCHEME "number:"
#define DIGITS_URI_SCHEME "digits:"
#define CHARACTERS_URI_SCHEME "characters:"
#define TONE_URI_SCHEME "tone:"
```

In the ARI source code.  Playing back sounds that exist on the asterisk server must be prefixed with "sound:" otherwise the schema error seen above occurs.

**Comment by shawty** [comment]
Sound scheme error is in fact not an error.  Sound uri's sent to asterisk need to be always qualified with a scheme in the Ubuntu 18.04 version.

The asterisk source code shows:

```
#define SOUND_URI_SCHEME "sound:"
#define RECORDING_URI_SCHEME "recording:"
#define NUMBER_URI_SCHEME "number:"
#define DIGITS_URI_SCHEME "digits:"
#define CHARACTERS_URI_SCHEME "characters:"
#define TONE_URI_SCHEME "tone:"
```

In the ARI source code.  Playing back sounds that exist on the asterisk server must be prefixed with "sound:" otherwise the schema error seen above occurs.

**Comment by skrusty** [comment]
Going to take look at this, this week. Will keep you posted.

**Comment by shawty** [comment]
No problems Ben, I'll not get back to my asterisk stuff until next weekend anyway, busy week coming up. :-)

That said, I can probably squeeze in a test or two here and there, so if you need to confirm anything shoot me a msg.

Peter

**Comment by nikriaz** [comment]
I spent a day trying to overcome WebSocket4Net, Version=0.15.0.0 issue. Nothing of standard solutions works, including re-bending dependency version. I strongly believe that AsterNET.ARI.dll (that is caller), calls WebSocket4Net in some incorrect, shortcut way, that is why <bindingRedirect oldVersion...newVersion/> it does not work. 

I did some play with dotPeek and found incorrect reference but have no idea for to do with it. 

![2019-06-22 21_29_55-JetBrains dotPeek](https://user-images.githubusercontent.com/52108512/59966408-d58b6b00-9535-11e9-8d46-2e0d7f9751f8.png)

**Comment by nikriaz** [comment]
OK, I was able to re-compile AsterNET.ARI.dll by editing its manifest to match current versions (as of today) and public keys of the dependencies. The problem was in  WebSocket4Net, SuperSocket.ClientEngine and Newtonsoft.JSON. Now it works but still it is something fundamentally wrong what strictly prevents to update dependencies.


[AsterNET.ARI.dll.zip](https://github.com/skrusty/AsterNET.ARI/files/3317228/AsterNET.ARI.dll.zip)

**Comment by nikriaz** [comment]
It's turned out that author already corrected these errors. Just download the latest source from this page and build your own dll in VS. The version should be 1.2.2 instead of 1.2.1 from nuget, which contains dependencies version mismatch.

**Comment by skrusty** [comment]
Thanks for the feedback, i will review this and update nuget as required.

**Comment by vdasus** [comment]
some news about nuget?

**Comment by shawty** [comment]
Actually I've been busy, forgot to follow this up.  As far as I'm aware the NuGet package has been updated, but I could be wrong, we'll wait to see if @skrusty replies.

**Comment by vdasus** [comment]
After updating nuget packages in solution - version 1.2.1

**Comment by shawty** [comment]
I'll test it later on, and if all's ok with my previous test I'll close this issue.

**Comment by vdasus** [comment]
Is there any news?

**Comment by shawty** [comment]
Ok, I've just gone back and tested my previous test app as detailed at the top of this issue.

It's configured as a full DotNet 4.7.1 app (Not .NET core), I've attached a zip of the project to this note.

I configured asterisk as above with an application called "pork" and the appropriate usernames/passwords etc and everything now appears to work.  I even managed to originate a real sip call from a real voip phone sat on my desk and that was handled perfectly by the AST Ari NuGet package.

As far as I can see, this is now fixed :-)

PS: I also updated to the latest versions of websocket and the other dependent libraries and everything seems to be perfectly fine.

[shawtyplay.zip](https://github.com/skrusty/AsterNET.ARI/files/3627437/shawtyplay.zip)
